src/lib/meta/meta_blk_service.cpp

```diff
@@ -1074,85 +1077,125 @@ sisl::byte_array MetaBlkService::read_sub_sb_internal(const meta_blk* mblk) cons
+
+                        cb(mblk, decompressed_buf, mblk->hdr.h.src_context_sz);
                     } else {
-                        // decompressed_size must equal to input sz before compress
-                        HS_REL_ASSERT_EQ(uint64_cast(mblk->hdr.h.src_context_sz),
-                                         uint64_cast(decompressed_size)); /* since decompressed_size is >=0 it
-                                                                             is safe to cast to uint64_t */
-                        HS_LOG(DEBUG, metablk,
-                               "[type={}] Successfully decompressed, compressed_sz: {}, src_context_sz: {}, "
-                               "decompressed_size: {}",
-                               mblk->hdr.h.type, uint64_cast(mblk->hdr.h.compressed_sz),
-                               uint64_cast(mblk->hdr.h.src_context_sz), decompressed_size);
+                        // There is use case that cb could be nullptr because client want to get its superblock via
+                        // read api;
+                        cb(mblk, buf, mblk->hdr.h.context_sz);
                     }
 
-                    cb(mblk, decompressed_buf, mblk->hdr.h.src_context_sz);
-                } else {
-                    // There is use case that cb could be nullptr because client want to get its superblock via
-                    // read api;
-                    cb(mblk, buf, mblk->hdr.h.context_sz);
+                    HS_LOG(DEBUG, metablk, "[type={}] meta blk sent with size: {}.", mblk->hdr.h.type,
+                           uint64_cast(mblk->hdr.h.context_sz));
                 }
-
-                HS_LOG(DEBUG, metablk, "[type={}] meta blk sent with size: {}.", mblk->hdr.h.type,
-                       uint64_cast(mblk->hdr.h.context_sz));
-            }
-        } else {
-            HS_LOG(DEBUG, metablk, "[type={}], unregistered client found. ");
+            } else {
+                HS_LOG(DEBUG, metablk, "[type={}], unregistered client found. ");
 #if 0
             // should never arrive here since we do assert on type before write to disk;
             HS_LOG_ASSERT( false, "[type={}] not registered for mblk found on disk. Skip this meta blk. ",
                       mblk->hdr.h.type);
 #endif
+            }
+        }
+
+        if (do_comp_cb && m_sub_info[subtype].comp_cb) {
+            m_sub_info[subtype].comp_cb(true);
+            HS_LOG(DEBUG, metablk, "[type={}] completion callback sent.", subtype);
         }
     }
+}
 
-    if (do_comp_cb) {
-        // for each registered subsystem, do recovery complete callback;
-        for (auto& sub : m_sub_info) {
-            if (sub.second.comp_cb) {
-                sub.second.comp_cb(true);
-                HS_LOG(DEBUG, metablk, "[type={}] completion callback sent.", sub.first);
-            }
+std::vector< meta_sub_type >
+MetaBlkService::topological_sort(std::unordered_map< meta_sub_type, std::vector< meta_sub_type > > topological_graph) {
```

Comment:
it will be better to use reference&. LG to me for the rest part.

Reply:
ack